cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cube/DefaultCube.java

```diff
@@ -113,8 +122,18 @@ public void add(Collection<? extends CubeFact> facts) {
       }
     }
 
-    for (FactTable table : resolutionToFactTable.values()) {
-      table.add(toWrite);
+    Map<Integer, Future> futures = new HashMap<>();
+    for (Map.Entry<Integer, FactTable> table : resolutionToFactTable.entrySet()) {
+      futures.put(table.getKey(), executorService.submit(() -> table.getValue().add(toWrite)));
+    }
+
+    for (Map.Entry<Integer, Future> future : futures.entrySet()) {
+      try {
+        future.getValue().get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new RuntimeException(String.format("Failed to add metrics to the %d " +
```

Comment:
If you throw in here, there could be other `Future` that are failed without one notice.

Reply:
Fixed.